packages/testing/src/chai-retry-plugin/chai-retry-plugin.ts

```diff
@@ -0,0 +1,187 @@
+                if (isNegate) {
+                    negationIsApplied = true;
+                    continue;
+                }
+
+                try {
+                    if (key) {
+                        assertion.to.be[key as keyof AssertionPropertyKeys];
+                    }
+
+                    method?.apply(assertion, args);
+                } catch (error) {
+                    if (error instanceof AssertionError && negationIsApplied) {
+                        continue;
+                    }
+
+                    throw error;
+                }
+
+                if (negationIsApplied) {
+                    throw new Error('Negated assertion should throw an error, but it finished successfully.');
+                }
+            }
+
+            return;
+        } catch (error) {
+            await sleep(delay);
+        }
+    }
+
+    throw new Error(`Limit of ${retries} retries exceeded!`);
+};
+
+const getRetryPromiseWithTimeout = (retryAndAssertProps: RetryAndAssertProps): Promise<void> => {
+    return addTimeoutToPromise(retryFunctionAndAssertResult(retryAndAssertProps), retryAndAssertProps.options.timeout);
+};
+
+/**
+ * Adds the `retry` method to Chai assertions, which allows to check the return value of a function until it satisfies the chained assertions.
+ * Should be applied through `Chai.use` function, for example:
+ * ```ts
+ * import Chai from 'chai';
+ * import { chaiRetryPlugin } from '@wixc3/testing';
+ *
+ * Chai.use(chaiRetryPlugin);
+ * ```
+ *
+ * Examples of usage:
+ * @example
+ * ```ts
+ * await expect(funcToRetry).retry({ timeout: 7000, retries: 5 }).have.property('value').and.be.above(4);
+ * ```
+ * @example
+ * ```ts
+ * await expect(sometimesNullFunction).retry({ timeout: 2000 }).to.be.not.null;
+ * ```
+ * @example
+ * ```ts
+ * await expect(funcToRetry).retry({ retries: 5, delay: 10 }).and.have.property('success').and.be.true;
+ * ```
+ */
+export const chaiRetryPlugin = function (_: typeof Chai, utils: Chai.ChaiUtils) {
+    Assertion.addMethod('retry', function (retryOptions: RetryOptions = {}): PromiseLikeAssertion {
+        const functionToRetry: FunctionToRetry = this._obj as FunctionToRetry;
+
+        if (typeof functionToRetry !== 'function') {
+            throw new TypeError(utils.inspect(functionToRetry) + ' is not a function.');
+        }
+
+        const assertionStack: AssertionStackItem[] = [];
+        const defaultRetryOptions: Required<RetryOptions> = { timeout: 5000, retries: Infinity, delay: 0 };
+        const options: Required<RetryOptions> = { ...defaultRetryOptions, ...retryOptions };
+
+        // Fake assertion object for catching calls of chained methods
+        const proxyTarget = new Assertion({});
+
+        /**
+         * This is needed to handle cases when assertion ends with property, for example:
+         * await expect(func).retry().to.be.null;
+         */
+        const assertionPropertyKeys = [
```

Comment:
why do we have this and `AssertionPropertyKeys` type.. why do we need both of them?

Reply:
@yurii-ve fixed